File: src/context_compiler/bundle.py
"""One exact rendered context for one computation.

Minimal extraction of the historical bundle record. A ContextBundle is
one exact ordered rendering: not a candidate pool, not a set, not a
retrieval result. `layout_trace` must equal item order exactly, and the
content hash is order-sensitive.

Provenance is an opaque passthrough (`dict | None`): the compiler
never sets it, never reads it, and round-trips whatever it was given
so historical records keep parsing byte-identically. Transport- or
capture-specific provenance belongs to downstream adapters, not to
the compiler domain.
"""

from __future__ import annotations

import hashlib
import uuid
from dataclasses import dataclass
from typing import Any

from context_compiler.items import SCHEMA_VERSION as ITEM_SCHEMA
from context_compiler.items import ContextItem

SCHEMA_VERSION = "project_context.context_bundle.v1"

__all__ = ["SCHEMA_VERSION", "ITEM_SCHEMA_VERSION", "ContextBundle", "build_bundle"]


@dataclass(frozen=True)
class ContextBundle:
    id: str
    items: tuple[ContextItem, ...]
    created_at: str
    layout_trace: tuple[str, ...]
    evidence_class: str = "synthetic"
    provenance: dict[str, Any] | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": SCHEMA_VERSION,
            "id": self.id,
            "items": [item.to_dict() for item in self.items],
            "created_at": self.created_at,
            "layout_trace": list(self.layout_trace),
            "evidence_class": self.evidence_class,
            "provenance": dict(self.provenance) if self.provenance is not None else None,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "ContextBundle":
        version = data.get("schema_version", SCHEMA_VERSION)
        if version != SCHEMA_VERSION:
            raise ValueError(f"unsupported ContextBundle schema: {version!r}")
        items = tuple(ContextItem.from_dict(raw) for raw in data["items"])
        provenance = data.get("provenance")
        if provenance is not None and not isinstance(provenance, dict):
            raise ValueError(f"invalid bundle provenance: {provenance!r}")
        bundle = cls(
            id=data["id"],
            items=items,
            created_at=data["created_at"],
            layout_trace=tuple(data["layout_trace"]),
            evidence_class=data.get("evidence_class", "synthetic"),
            provenance=dict(provenance) if provenance is not None else None,
        )
        bundle._check_layout()
        return bundle

    def _check_layout(self) -> None:
        rendered = [item.id for item in self.items]
        if list(self.layout_trace) != rendered:
            raise ValueError("layout_trace must equal item order exactly")

    def rendered_token_total(self) -> int:
        return sum(item.token_count for item in self.items)

    def content_hash(self) -> str:
        """Order-sensitive identity over rendered content. [A,B,C] and
        [C,B,A] hash differently even with identical membership."""
        digest = hashlib.sha256()
        for item in self.items:
            digest.update(item.id.encode("utf-8"))
            digest.update(b"\x00")
            digest.update(item.content.encode("utf-8"))
            digest.update(b"\x00")
        return digest.hexdigest()


def build_bundle(
    items: list[ContextItem],
    *,
    bundle_id: str | None = None,
    created_at: str,
    evidence_class: str = "synthetic",
    provenance: dict[str, Any] | None = None,
) -> ContextBundle:
    """Assign positions 0..n in given order and freeze. Deterministic
    callers pass stable ids; a UUID is assigned only when omitted."""
    ordered = tuple(
        ContextItem(
            id=item.id,
            source=item.source,
            kind=item.kind,
            content=item.content,
            position=index,
            token_count=item.token_count,
            token_provenance=item.token_provenance,
            authority=item.authority,
            scope=item.scope,
            observed_at=item.observed_at,
            semantic_id=item.semantic_id,
            ref=item.ref,
        )
        for index, item in enumerate(items)
    )
    bundle = ContextBundle(
        id=bundle_id or f"bundle-{uuid.uuid4().hex[:12]}",
        items=ordered,
        created_at=created_at,
        layout_trace=tuple(item.id for item in ordered),
        evidence_class=evidence_class,
        provenance=provenance,
    )
    bundle._check_layout()
    return bundle


ITEM_SCHEMA_VERSION = ITEM_SCHEMA

File: src/core/bundle.ts
/**
 * ContextBundle record (v1 schema preserved). One exact ordered
 * rendering for one computation: layout_trace equals item order
 * exactly, and the content hash is order-sensitive.
 *
 * Hashing replicates the Python reference byte-for-byte:
 * UTF-8(item.id) + NUL + UTF-8(item.content) + NUL per item in
 * order, then SHA-256. Provenance is an opaque passthrough.
 */

import { createHash, randomUUID } from "node:crypto";
import { itemFromJSON, itemToJSON, type ContextItem } from "./items.ts";

export const BUNDLE_SCHEMA = "project_context.context_bundle.v1";
export const ITEM_SCHEMA_VERSION = "project_context.context_item.v1";

export interface ContextBundle {
  readonly id: string;
  readonly items: readonly ContextItem[];
  readonly createdAt: string;
  readonly layoutTrace: readonly string[];
  readonly evidenceClass: string;
  readonly provenance: Record<string, unknown> | null;
}

export function bundleToJSON(bundle: ContextBundle): Record<string, unknown> {
  return {
    schema_version: BUNDLE_SCHEMA,
    id: bundle.id,
    items: bundle.items.map(itemToJSON),
    created_at: bundle.createdAt,
    layout_trace: [...bundle.layoutTrace],
    evidence_class: bundle.evidenceClass,
    provenance: bundle.provenance ? { ...bundle.provenance } : null,
  };
}

export function bundleFromJSON(data: unknown): ContextBundle {
  if (typeof data !== "object" || data === null) {
    throw new Error("bundle is not an object");
  }
  const raw = data as Record<string, unknown>;
  const version = (raw["schema_version"] ?? BUNDLE_SCHEMA) as string;
  if (version !== BUNDLE_SCHEMA) {
    throw new Error(
      `unsupported ContextBundle schema: ${JSON.stringify(version)}`,
    );
  }
  const provenance = raw["provenance"] ?? null;
  if (provenance !== null && typeof provenance !== "object") {
    throw new Error("invalid bundle provenance");
  }
  const itemsRaw = raw["items"];
  if (!Array.isArray(itemsRaw)) throw new Error("invalid bundle items");
  const traceRaw = raw["layout_trace"];
  if (
    !Array.isArray(traceRaw) ||
    !traceRaw.every((v) => typeof v === "string")
  ) {
    throw new Error("invalid layout_trace");
  }
  const bundle: ContextBundle = {
    id: raw["id"] as string,
    items: itemsRaw.map(itemFromJSON),
    createdAt: raw["created_at"] as string,
    layoutTrace: [...(traceRaw as string[])],
    evidenceClass: (raw["evidence_class"] ?? "synthetic") as string,
    provenance: provenance
      ? { ...(provenance as Record<string, unknown>) }
      : null,
  };
  checkLayout(bundle);
  return bundle;
}

export function checkLayout(bundle: ContextBundle): void {
  const rendered = bundle.items.map((item) => item.id);
  if (
    rendered.length !== bundle.layoutTrace.length ||
    rendered.some((id, index) => id !== bundle.layoutTrace[index])
  ) {
    throw new Error("layout_trace must equal item order exactly");
  }
}

export function renderedTokenTotal(bundle: ContextBundle): number {
  return bundle.items.reduce((sum, item) => sum + item.tokenCount, 0);
}

export function contentHash(bundle: ContextBundle): string {
  const digest = createHash("sha256");
  for (const item of bundle.items) {
    digest.update(item.id, "utf-8");
    digest.update(Buffer.from([0]));
    digest.update(item.content, "utf-8");
    digest.update(Buffer.from([0]));
  }
  return digest.digest("hex");
}

export function buildBundle(input: {
  items: readonly ContextItem[];
  bundleId?: string | null;
  createdAt: string;
  evidenceClass?: string;
  provenance?: Record<string, unknown> | null;
}): ContextBundle {
  const ordered = input.items.map((item, index) => ({
    ...item,
    position: index,
  }));
  const bundle: ContextBundle = {
    id:
      input.bundleId ?? `bundle-${randomUUID().replace(/-/g, "").slice(0, 12)}`,
    items: ordered,
    createdAt: input.createdAt,
    layoutTrace: ordered.map((item) => item.id),
    evidenceClass: input.evidenceClass ?? "synthetic",
    provenance: input.provenance ?? null,
  };
  checkLayout(bundle);
  return bundle;
}
